src/lib/shadow-shim-helper-rs/src/rootedcell/cell.rs

```diff
@@ -0,0 +1,163 @@
+use super::{Root, Tag};
+use std::cell::UnsafeCell;
+use vasi::VirtualAddressSpaceIndependent;
+
+/// Analagous to [std::cell::Cell]. In particular like [std::cell::Cell], it
+/// doesn't perform any atomic operations internally, making it relatively
+/// inexpensive.
+///
+/// Unlike [std::cell::Cell], this type is [Send] and [Sync] if `T` is
+/// [Send]. This is safe because the owner is required to prove access to the
+/// associated [Root], which is `![Sync]`, to access.
+#[derive(Debug)]
+#[repr(C)]
+pub struct RootedCell<T>
+where
+    T: Copy,
+{
+    tag: Tag,
+    val: UnsafeCell<T>,
+}
+
+// SAFETY: RootedCell is VirtualAddressSpaceIndependent as long as T is.
+unsafe impl<T> VirtualAddressSpaceIndependent for RootedCell<T> where
+    T: VirtualAddressSpaceIndependent + Copy
+{
+}
+
+impl<T> RootedCell<T>
+where
+    T: Copy,
+{
+    /// Create a RootedCell associated with `root`.
+    pub fn new(root: &Root, val: T) -> Self {
+        Self {
+            tag: root.tag(),
+            val: UnsafeCell::new(val),
+        }
+    }
+
+    pub fn get(&self, root: &Root) -> T {
+        // Prove that the root is held for this tag.
+        assert_eq!(
+            root.tag, self.tag,
+            "Expected {:?} Got {:?}",
+            self.tag, root.tag
+        );
+
+        unsafe { *self.val.get() }
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        // Since we have the only reference to `self`, we don't need to check the root.
+        unsafe { &mut *self.val.get() }
+    }
+
+    pub fn set(&self, root: &Root, val: T) {
+        // Prove that the root is held for this tag.
+        assert_eq!(
+            root.tag, self.tag,
+            "Expected {:?} Got {:?}",
+            self.tag, root.tag
+        );
+
+        unsafe { *self.val.get() = val }
+    }
+
+    pub fn replace(&self, root: &Root, val: T) -> T {
+        // Prove that the root is held for this tag.
+        assert_eq!(
+            root.tag, self.tag,
+            "Expected {:?} Got {:?}",
+            self.tag, root.tag
+        );
+
+        let old = unsafe { *self.val.get() };
+        unsafe { *self.val.get() = val }
+        old
```

Comment:
optional nit: Maybe consider using the pointer module functions instead, just because they're a little easier to reason about. For example sometimes dereferencing a pointer in rust actually causes it to create a reference. I wouldn't expect it to in this particular case, but I find dereferencing raw pointers to be confusing in rust because of it. For example `*x += 5;` where `x` is a pointer creates an implicit mutable reference to the integer (so it can be passed to the `AddAssign` trait), which is unexpected and may cause an aliasing issue if you already have another mutable reference to that integer.

The dereference-assignment here is fine I think, but it might be nicer to do a `std::ptr::replace` here anyways.